src/librustc/hir/pat_util.rs

```diff
@@ -77,6 +77,34 @@ impl hir::Pat {
         });
     }
 
+    /// Call `f` on every "binding" in a pattern, e.g., on `a` in
+    /// `match foo() { Some(a) => (), None => () }`.
+    ///
+    /// When encountering an or-pattern `p_0 | ... | p_n` only `p_0` will be visited.
+    pub fn each_binding_or_first<F>(&self, c: &mut F)
```

Comment:
```suggestion
    pub fn each_binding_or_first<F>(&self, f: &mut F)
```

Reply:
Is calling `each_binding_or_first` recursively going to have the right behaviour for nested `Or` patterns? Shouldn't it pick the first pattern only at the top level? (In which case, you can use `each_binding` instead and just check the top level at the call site.)

Reply:
(Using `c` for callback as `f` is used for `FieldPat`; could switch to `g` for the callback instead.)

(We discussed this a bit privately on Discord; The existing commentary in `define_bindings_in_pat` states that we only consider the first pattern since any later patterns must have the same bindings (and the first pattern is considered to have the authoritative set of ids). This suggests that we should do the same in nested positions as there should be no intrinsic semantic difference between `Some(a @ 1) | Some(a @ 2)` and `Some(a @ (1 | 2))`. That said, I don't really know *why* we only consider the first pattern and if we just use `.each_binding(...)` uniformly for top/nested then the UI tests pass at least.)

Reply:
Looking at blame, it seems the comment was left by @eddyb 5 years ago. Maybe they still know. =P

Reply:
It's actually due to @nikomatsakis here: https://github.com/rust-lang/rust/commit/e47d2f60607142eaf68a0c560f5c3d37fd1da13d#diff-97c8d3b79a36724da452cf176a6a42b5R661-R663.

Reply:
I don't even understand what the question is here. :)

Reply:
@nikomatsakis So your comment to a call of `.define_bindings_in_pat(...)` in `liveness.rs` regarding `ExprKind::Match(...) =>` states that:

```rust
                    // only consider the first pattern; any later patterns must have
                    // the same bindings, and we also consider the first pattern to be
                    // the "authoritative" set of ids
                    let arm_succ =
                        self.define_bindings_in_arm_pats(arm.pats.first().map(|p| &**p),
                                                         guard_succ);
```

Now, I don't exactly understand *why* this is, but my intuition says that if we want to extend this logic to nested or-patterns, e.g. `Some(A(x) | B(x))` then we should do the same think there. That is, in any `PatKind::Or(pats)` we will just visit `pats[0]` and not `pats[1..]`. That's what the code in this PR does now.

Reply:
Well, so here's the problem. Presently we use the id of the AST node for a pattern binding as the variable's id -- but when you have e.g. `Foo(x) | Bar(x)`, you wind up with two id's for `x`! Clearly all the alternatives must have the same *set* of bindings (i.e., they must all define `x`), so yeah I dimly recall deciding that we would just use the id's from the *first* set of patterns to define the variables, and then check the remaining patterns against that first set. 

I think another viable approach would be to use some kind of fresh identifiers for the variables (not the id from the pattern node) and then be able to "resolve" each pattern node to that. Or perhaps to have a table of "redirects" where for most bindings the redirect is the identity, but for `|` patterns they point to the bindings from the first set. 

In any case I think you are correct that if we are moving `|` patterns further down, and we want to try and use the same strategy, then we only want to take the left-most branches (but check the other branches for consistency).

Reply:
Thanks for the elaboration!

> Clearly all the alternatives must have the same _set_ of bindings (i.e., they must all define `x`), so yeah I dimly recall deciding that we would just use the id's from the _first_ set of patterns to define the variables, and then check the remaining patterns against that first set.

> (but check the other branches for consistency).

Yea, that's what we do; I recently tweaked this logic in resolve here: https://github.com/rust-lang/rust/blob/9b9d2aff8de4d499b4ba7ca406e000f8d3754ea7/src/librustc_resolve/late.rs#L1121-L1475

> I think another viable approach would be to use some kind of fresh identifiers for the variables (not the id from the pattern node) and then be able to "resolve" each pattern node to that. Or perhaps to have a table of "redirects" where for most bindings the redirect is the identity, but for `|` patterns they point to the bindings from the first set.

Sounds interesting (but should be done in a follow-up if so). cc @petrochenkov may be interested.